[labeling] specify unit for data defined label rotation

domi4484 opened the pull request:
Specify the angle unit for data defined label rotation

![image](https://user-images.githubusercontent.com/9881900/126294292-06c91bd5-83d2-4d6f-88a2-38fa2d303c86.png)

![Peek 2021-08-29 14-38](https://user-images.githubusercontent.com/9881900/131250678-52360214-31a9-4cf9-ad84-b10a4f78c431.gif)

domi4484 committed:
[labeling] specify unit for data defined label rotation

nyalldawson commented on src/core/labeling/qgspallabeling.h:
```diff
@@ -636,6 +636,9 @@ class CORE_EXPORT QgsPalLayerSettings
     bool preserveRotation = true;
 
+    //! Unit for rotation of labels.
+    QgsUnitTypes::AngleUnit rotationUnit = QgsUnitTypes::AngleDegrees;
```
Can you make the member variable private and add getter/setter functions? Having public member variables makes things very tricky with the stable API restriction

domi4484 replied:
At line 18 it is stated that this file is not part of the stable API. Anyway I would have nothing against moving it to private but all other properties are in the public section too. For code consistency I would leave it there for now and add a TODO comment for QGIS 4 that all this members should become private. What do you think?

nyalldawson replied:
That's an extremely old comment which should definitely be removed -- this class has been stable since early 2.x. I've been slowly cleaning up this class but it's taken a lot of messy code because of the public members, which is why I don't want any more added.

nyalldawson reviewed (commented):
Got some unit tests coming?

domi4484 commented:
> Got some unit tests coming?

Would like but it is not clear to me how to implement such test... I have just found this [tests/testdata/labeling/README.rst](https://github.com/qgis/QGIS/blob/master/tests/testdata/labeling/README.rst), I'll gave a look and try to implement some tests.

nyalldawson commented:
> Would like but it is not clear to me how to implement such test... I have just found this tests/testdata/labeling/README.rst, I'll gave a look and try to implement some tests.

Those tests are pretty hard to work with. I'd suggest just adding a test to testqgslabelingengine.cpp instead.

3nids commented on src/core/labeling/qgspallabeling.cpp:
```diff
@@ -1024,6 +1026,7 @@ void QgsPalLayerSettings::readXml( const QDomElement &elem, const QgsReadWriteCo
   }
 
   preserveRotation = placementElem.attribute( QStringLiteral( "preserveRotation" ), QStringLiteral( "1" ) ).toInt();
+  rotationUnit = static_cast<QgsUnitTypes::AngleUnit>( placementElem.attribute( QStringLiteral( "rotationUnit" ), 0 ).toInt() );
```
what about saving this as a string and using:

```suggestion
  rotationUnit = qgsEnumKeyToValue( placementElem.attribute( QStringLiteral( "rotationUnit" ), qgsEnumValueToKey( QgsUnitTypes::AngleDegrees ) ), QgsUnitTypes::AngleDegrees );
```

3nids commented on src/core/labeling/qgspallabeling.cpp:
```diff
@@ -1213,6 +1216,7 @@ QDomElement QgsPalLayerSettings::writeXml( QDomDocument &doc, const QgsReadWrite
   placementElem.setAttribute( QStringLiteral( "labelOffsetMapUnitScale" ), QgsSymbolLayerUtils::encodeMapUnitScale( labelOffsetMapUnitScale ) );
   placementElem.setAttribute( QStringLiteral( "rotationAngle" ), angleOffset );
   placementElem.setAttribute( QStringLiteral( "preserveRotation" ), preserveRotation );
+  placementElem.setAttribute( QStringLiteral( "rotationUnit" ), static_cast< unsigned int >( rotationUnit ) );
```
```suggestion
  placementElem.setAttribute( QStringLiteral( "rotationUnit" ), qgsEnumValueToKey( rotationUnit ) );
```

3nids reviewed (commented):
small remarks regarding using enums

domi4484 commented on src/core/labeling/qgspallabeling.cpp:
```diff
@@ -797,6 +798,7 @@ void QgsPalLayerSettings::readFromLayerCustomProperties( QgsVectorLayer *layer )
   }
 
   preserveRotation = layer->customProperty( QStringLiteral( "labeling/preserveRotation" ), QVariant( true ) ).toBool();
+  rotationUnit = static_cast<QgsUnitTypes::AngleUnit>( layer->customProperty( QStringLiteral( "labeling/rotationUnit" ), QVariant( static_cast< unsigned int >( QgsUnitTypes::AngleDegrees ) ) ).toInt() );
```
Nice thank you!

3nids replied:
it's merged.

domi4484 committed:
[labeling] test for rotation unit

domi4484 committed:
fix test

domi4484 committed:
added mask image

nyalldawson commented on src/core/labeling/qgspallabeling.h:
```diff
@@ -636,6 +633,12 @@ class CORE_EXPORT QgsPalLayerSettings
     //! True if label rotation should be preserved during label pin/unpin operations.
     bool preserveRotation = true;
 
+    //! Unit for rotation of labels.
+    QgsUnitTypes::AngleUnit rotationUnit() const;
```
```suggestion
    /**
     * Unit for rotation of labels.
     * \see setRotationUnit()
     * \since QGIS 3.22
     */
    QgsUnitTypes::AngleUnit rotationUnit() const;
```

nyalldawson commented on src/core/labeling/qgspallabeling.h:
```diff
@@ -636,6 +633,12 @@ class CORE_EXPORT QgsPalLayerSettings
+    //! Unit for rotation of labels.
+    QgsUnitTypes::AngleUnit rotationUnit() const;
+
+    //! Set unit for rotation of labels.
```
```suggestion
    /**
      * Set unit for rotation of labels.
      * \see rotationUnit()
      * \since QGIS 3.22
      */
```

domi4484 committed:
Apply suggestions from code review

Co-authored-by: Nyall Dawson <nyall.dawson@gmail.com>

domi4484 committed:
QgsMapToolRotateLabel respect rotation unit

domi4484 committed:
Map rotate too display rotation unit

domi4484 commented:
@3nids @nyalldawson since you last viewed I added the unit in the "Rotate label tool" and fixed all your suggestions.
Could someone also remove the stale label please?

nyalldawson commented on src/core/qgsunittypes.cpp:
```diff
@@ -2572,6 +2572,52 @@ QString QgsUnitTypes::formatAngle( double angle, int decimals, QgsUnitTypes::Ang
+      break;
+  }
+
+  return QStringLiteral( "%L1%2" ).arg( angle, 0, 'f', decimals ).arg( unitLabel );
```
I think this has to use QLocale().toString now, not the direct QString approach. @elpaso can you confirm?

domi4484 replied:
@nyalldawson as I understand from the Qt documentation the `L` in the placeholder is the same as using QLocale().toString

nyalldawson commented on src/core/qgsunittypes.h:
```diff
@@ -563,6 +563,13 @@ class CORE_EXPORT QgsUnitTypes
+     * \param unit unit of angle
+     * \returns formatted angle string
+     */
+    Q_INVOKABLE static QString formatAngle( double angle, QgsUnitTypes::AngleUnit unit );
```
Could you add a "decimals" argument here, default to -1 and add a note that "-1 means automatically pick an appropriate number of decimal places for the unit". (also needs some unit tests)

domi4484 replied:
Thank you I'll implement the "-1" suggestion. There is already an overload which takes a decimals [parameter](https://github.com/qgis/QGIS/blob/83a441589a544015897640f6a8bd06043f59f9d6/src/core/qgsunittypes.h#L564), but it is the second param so I can't set the default value to -1. Or should I set a default parameter for the unit to? like `=AngleDegrees`?

domi4484 replied:
I have removed this function and modified the original one which now can take -1 for decimals. Unit tests where updated too.

nyalldawson commented on src/core/qgsunittypes.h:
```diff
@@ -563,6 +563,13 @@ class CORE_EXPORT QgsUnitTypes
+     * Returns an angle formatted as a friendly string.
+     * \param angle angle to format
+     * \param unit unit of angle
+     * \returns formatted angle string
```
```suggestion
     * \returns formatted angle string
     * \since QGIS 3.22
```

nyalldawson reviewed (commented):
Nearly there, thanks!

domi4484 committed:
QgsUnitTypes::FormatAngle show appropriate number of decimals

m-kuhn reviewed (approved):
Can we merge this tomorrow or are there still open questions?
It has been around for a while already :-)

nyalldawson commented on src/core/qgsunittypes.h:
```diff
@@ -557,13 +557,12 @@ class CORE_EXPORT QgsUnitTypes
      * Returns an angle formatted as a friendly string.
      * \param angle angle to format
-     * \param decimals number of decimal places to show
+     * \param decimals number of decimal places to show. By -1 it show an appropriate number of decimal places
```
```suggestion
     * \param decimals number of decimal places to show. A value of -1 indicates that an appropriate number of decimal places should automatically be selected.
```

domi4484 replied:
thank you @nyalldawson , i committed it right now

nyalldawson reviewed (commented):
One very minor last change please!

domi4484 committed:
Better decimals parameter description

m-kuhn commented:
Thanks a lot @nyalldawson
